plugins/auth-backend/src/providers/google/provider.ts

```diff
@@ -133,89 +129,130 @@ export class GoogleAuthProvider implements OAuthHandlers {
-        { token },
-      );
-
-      return {
-        ...response,
-        backstageIdentity: {
-          id: user.metadata.name,
+        {
+          tokenIssuer: this.tokenIssuer,
+          catalogIdentityClient: this.catalogIdentityClient,
         },
-      };
-    } catch (error) {
-      this.logger.warn(
-        `Failed to look up user, ${error}, falling back to allowing login based on email pattern, this will probably break in the future`,
       );
-      return {
-        ...response,
-        backstageIdentity: { id: profile.email.split('@')[0] },
-      };
     }
+
+    return response;
   }
 }
 
-export type GoogleProviderOptions = {};
+const emailSignInResolver: SignInResolver<OAuthResult> = async (info, ctx) => {
+  const { profile } = info;
+
+  if (!profile.email) {
+    throw new Error('Google profile contained no email');
+  }
+
+  const entity = await ctx.catalogIdentityClient.findUser({
+    annotations: {
+      'google.com/email': profile.email,
+    },
+  });
+
+  const claims = getEntityClaims(entity);
+  const token = await ctx.tokenIssuer.issueToken({ claims });
+
+  return { id: entity.metadata.name, entity, token };
+};
+
+export type GoogleProviderOptions = {
+  /**
+   * The profile transformation function used to verify and convert the auth response
+   * into the profile that will be presented to the user.
+   */
+  profileTransform?: ProfileTransform<OAuthResult>;
+
+  /**
+   * Configure sign-in for this provider, without it the provider can not be used to sign users in.
+   */
+  signIn?: {
+    /**
+     * Maps an auth result to a Backstage identity for the user.
+     *
+     * Set to `'email'` to use the default email-based sign in resolver, which will search
+     * for the catalog for a single user entity that has a matching `google.com/email` annotation.
+     */
+    resolver?: 'email' | SignInResolver<OAuthResult>;
+  };
+};
 
 export const createGoogleProvider = (
-  _options?: GoogleProviderOptions,
+  options?: GoogleProviderOptions,
 ): AuthProviderFactory => {
-  return ({
-    providerId,
-    globalConfig,
-    config,
-    logger,
-    tokenIssuer,
-    catalogApi,
-  }) =>
+  return ({ providerId, globalConfig, config, tokenIssuer, catalogApi }) =>
     OAuthEnvironmentHandler.mapConfig(config, envConfig => {
       const clientId = envConfig.getString('clientId');
       const clientSecret = envConfig.getString('clientSecret');
       const callbackUrl = `${globalConfig.baseUrl}/${providerId}/handler/frame`;
 
+      const catalogIdentityClient = new CatalogIdentityClient({
+        catalogApi,
+        tokenIssuer,
+      });
+
+      let profileTransform: ProfileTransform<OAuthResult> = async ({
+        fullProfile,
+        params,
+      }) => makeProfileInfo(fullProfile, params.id_token);
+      if (options?.profileTransform) {
+        profileTransform = options.profileTransform;
+      }
+
+      let signInResolver: SignInResolver<OAuthResult> | undefined = undefined;
```

Comment:
Should this instead default to `emailSignInResolver`, similar to `profileTransform` above? The special `email` key seems unnecessary for an optional field that could fallback to the this default.

If I make a factory defining only a custom `profileTransform` and forget to set the optional `signIn.resolver: email` I unexpectedly lose the default behavior here.